ext/POSIX/POSIX.xs

```diff
@@ -3348,30 +3346,49 @@ write(fd, buffer, nbytes)
 void
 abort()
 
-#ifdef I_WCHAR
-#  include <wchar.h>
+#if defined(HAS_MBRLEN) && (defined(USE_ITHREADS) || ! defined(HAS_MBLEN))
+#  define USE_MBRLEN
+#else
+#  undef USE_MBRLEN
 #endif
 
 int
 mblen(s, n)
-	char *		s
+	SV *		s
 	size_t		n
-    PREINIT:
-#if defined(USE_ITHREADS) && defined(HAS_MBRLEN)
-        mbstate_t ps;
-#endif
     CODE:
-#if defined(USE_ITHREADS) && defined(HAS_MBRLEN)
-        memset(&ps, 0, sizeof(ps)); /* Initialize state */
-        RETVAL = mbrlen(s, n, &ps); /* Prefer reentrant version */
-#else
-        /* This might prevent some races, but locales can be switched out
-         * without locking, so this isn't a cure all */
-        LOCALE_LOCK;
-
-        RETVAL = mblen(s, n);
-        LOCALE_UNLOCK;
+        errno = 0;
+        if (! SvOK(s)) {
+#ifdef USE_MBRLEN
+            /* Initialize the shift state in PL_mbrlen_ps.  It's a good guess
+             * that the initial state is all zeros.  On some platforms, it is
+             * documented that pointing the first parameter to L'\0'
+             * initializes the state; on others, that doesn't work, but making
+             * the parameter NULL does.  Perhaps NULL works on all platforms,
+             * but this seems safer. */
```

Comment:
This is (badly) documented in the C standard as

> A zero-valued mbstate_t object is (at least) one way to describe an initial conversion state. A zerovalued
mbstate_t object can be used to initiate conversion involving any multibyte character sequence, in any LC_CTYPE category setting.

but doesn't define what "zero-valued" is for what is to us an opaque object.  It's up to you whether this comment should reflect that